### PR TITLE
Deprecate complex objects as an attributes `defaultValue`.

### LINKED ISSUE #3766: [Request] deprecate `defaultValue: primitiveValue`
https://github.com/emberjs/data/blob/c208f2c299ecd5566ad9d0eee0a89f9913e924c6/packages/ember-data/lib/system/model/attributes.js#L231

It is quite common to see: `defaultValue: []` but this results in a shared array between all instances, obviously nearly never something a user actual intends.

An alternative, and likely the better option, is to only allow primitive values, and require functions for complex objects. If someone wants the old behavior, that can close over a shared complex object.

### PR DESCRIPTION
**Pull request by bmac**
They often do not work how a user expects.
Closes #3766

### CODE REVIEW

**Comment by wecc** [comment]
:clap:

File: packages/ember-data/lib/system/model/attributes.js [hunk 1]
```diff
@@ -226,7 +226,13 @@ function getDefaultValue(record, options, key) {
   if (typeof options.defaultValue === "function") {
     return options.defaultValue.apply(null, arguments);
   } else {
-    return options.defaultValue;
+    let defaultValue = options.defaultValue;
+    Ember.deprecate(`Non primative defaultValues are deprecated because they are shared between all instances. If you would like to use a complex object as a default value please provide a function that returns the complex object.`,
```
Comment (sly7-7) [inline packages/ember-data/lib/system/model/attributes.js]: small typo 'primitive'

File: packages/ember-data/tests/unit/model-test.js [hunk 2]
```diff
@@ -611,6 +611,41 @@ test("a defaultValue function gets the record, options, and key", function() {
...
+  var store = createStore({
+    tag: Tag
+  });
+
+  run(function() {
+    tag = store.createRecord('tag');
+  });
+  expectDeprecation(function() {
+    get(tag, 'tagInfo');
+  }, /Non primative defaultValues are deprecated/);
```
Comment (sly7-7) [inline packages/ember-data/tests/unit/model-test.js]: same here ;)

**Comment by igorT** [comment]
:heart:

**Pull request merged**